File: 프로그래머스/3/67259. ［카카오 인턴］ 경주로 건설/［카카오 인턴］ 경주로 건설.py
import heapq

def solution(board):
    n = len(board)
    costs = [[{'v':float('inf'), 'h':float('inf')} for _ in range(n)] for _ in range(n)]
    costs[0][0]['v'] = 0
    costs[0][0]['h'] = 0
    dir = [(1,0,'h'), (-1,0,'h'), (0,1,'v'), (0,-1,'v')]
    queue = [(0, 0, 0, None)]
    
    while queue:
        cur_cost, x, y, d = heapq.heappop(queue)
        
        for dx, dy, nd in dir:
            nx, ny = x+dx, y+dy
            if nx<0 or nx>=n or ny<0 or ny>=n or board[nx][ny]==1: continue
            next_cost = cur_cost+600 if d and d!=nd else cur_cost + 100
            if next_cost<costs[nx][ny][nd]:
                costs[nx][ny][nd] = next_cost
                heapq.heappush(queue, (next_cost, nx, ny, nd))
                
    return min(costs[n-1][n-1]['h'], costs[n-1][n-1]['v'])

File: 프로그래머스/3/67259. ［카카오 인턴］ 경주로 건설/［카카오 인턴］ 경주로 건설.java
import java.util.*;

class Solution {
    static class Node {
        int x;
        int y;
        int cost;
        String dir;

        public Node(int x, int y, int cost, String dir) {
            this.x = x;
            this.y = y;
            this.cost = cost;
            this.dir = dir;
        }
    }

    static class Cost {
        int g;
        int s;

        public Cost(int g, int s) {
            this.g = g;
            this.s = s;
        }
    }

    public int solution(int[][] board) {
        int N = board.length;
        PriorityQueue<Node> pq = new PriorityQueue<>((o1, o2) -> Integer.compare(o1.cost, o2.cost));
        Node start = new Node(0, 0, 0, null);
        pq.offer(start);
        int[][] dirs = new int[][] {{0,1}, {0,-1}, {1,0}, {-1,0}};
        Cost[][] costs = new Cost[N][N];
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                costs[i][j] = new Cost(Integer.MAX_VALUE,Integer.MAX_VALUE);
            }
        }
        costs[0][0].g = 0;
        costs[0][0].s = 0;

        while (!pq.isEmpty()) {
            Node q = pq.poll();
            int cost = q.cost;

            for (int[] dir : dirs) {
                int dx = q.x + dir[0];
                int dy = q.y + dir[1];

                String d;
                if (dx==q.x) {
                    d = "g";
                } else {
                    d = "s";
                }

                int nextCost = cost;
                if (q.dir != null && !q.dir.equals(d)) {
                    nextCost += 600;
                } else {
                    nextCost += 100;
                }

                if (dx<0 || dx>=N || dy<0 || dy>=N || board[dx][dy] == 1) continue;
                if ((d.equals("g")&&costs[dx][dy].g <= nextCost) || (d.equals("s")&&costs[dx][dy].s<=nextCost)) continue;

                if (d.equals("g")) costs[dx][dy].g = nextCost;
                else costs[dx][dy].s = nextCost;
                pq.offer(new Node(dx, dy, nextCost, d));
            }
        }

        return Math.min(costs[N-1][N-1].s, costs[N-1][N-1].g);
    }
}